### PR TITLE
update s3 etl lambdafied

### PR DESCRIPTION
**Pull request by Kurt812**
Adapted the etl script to sit on an AWS lambda.
Created the lambda and related resources using terraform.
Script now runs on lambda, when invoked, to update the keyword_recordings archive file.
Next steps create a scheduler to trigger the lambda.

### CODE REVIEW

File: rds_to_s3/etl_lambda.py [hunk 1]
```diff
@@ -11,55 +11,38 @@
 import boto3
 from botocore.exceptions import NoCredentialsError, ClientError
 
+load_dotenv()
```
Comment (SurinaCS) [inline rds_to_s3/etl_lambda.py:14]: nice call

File: rds_to_s3/etl_lambda.py [hunk 2]
```diff
@@ -11,55 +11,38 @@
...
         FROM
             {SCHEMA_NAME}.keyword_recordings kr
-        ORDER BY kr.keyword_recordings_id ASC;
+        WHERE created_at < NOW() - INTERVAL '24 HOURS'
         """
-]
+REMOVE_QUERY = f"""
+        DELETE FROM {SCHEMA_NAME}.keyword_recordings
+        WHERE created_at < NOW() - INTERVAL '24 HOURS'
+    """
```
Comment (SurinaCS) [inline rds_to_s3/etl_lambda.py:27]: nice separation of concerns

File: rds_to_s3/etl_lambda.py [hunk 3]
```diff
@@ -70,8 +53,7 @@ def setup_connection() -> None:
             database=ENV["DB_NAME"]
         )
         cursor = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
-        cursor.execute(f"SET SEARCH_PATH TO {ENV['SCHEMA_NAME']};")
-        return conn
+        return conn, cursor
     except OperationalError as oe:
         logging.error(
             "Operational error while connecting to the database: %s", oe)
```
Comment (SurinaCS) [inline rds_to_s3/etl_lambda.py:59]: great error handling

File: rds_to_s3/etl_lambda.py [hunk 4]
```diff
@@ -93,39 +75,41 @@ def s3_connection() -> boto3.client:
...
         return pd.DataFrame()
     except pd.errors.EmptyDataError as ede:
-        logging.warning("Downloaded file %s is empty: %s", file_name, ede)
+        logging.warning("Downloaded file %s is empty: %s", tmp_file_name, ede)
         return pd.DataFrame()
     except Exception as e:
         logging.error(
-            "Unexpected error while downloading %s from S3: %s", file_name, e)
+            "Unexpected error while downloading %s from S3: %s", tmp_file_name, e)
         raise
```
Comment (SurinaCS) [inline rds_to_s3/etl_lambda.py:104]: why is the exception the only one not returning a dataframe? should they all raise their own error to interrupt the flow?

File: terraform/pipeline_ecs.tf [hunk 5]
```diff
@@ -97,8 +97,8 @@ resource "aws_ecs_task_definition" "c14_trendgineers_pipeline" {
         { name = "SCHEMA_NAME", value = var.SCHEMA_NAME },
         { name = "DB_NAME", value = var.DB_NAME },
         { name = "DB_PASSWORD", value = var.DB_PASSWORD },
-        { name = "AWS_ACCESS_KEY_ID", value = var.AWS_ACCESS_KEY_ID },
-        { name = "AWS_SECRET_ACCESS_KEY", value = var.AWS_SECRET_ACCESS_KEY }
+        { name = "AWS_ACCESS_KEY_ID", value = var.ACCESS_KEY_ID },
+        { name = "AWS_SECRET_ACCESS_KEY", value = var.SECRET_ACCESS_KEY }
```
Comment (SurinaCS) [inline terraform/pipeline_ecs.tf:101]: nice updating

File: terraform/pipeline_ecs.tf [hunk 6]
```diff
@@ -149,18 +149,18 @@ resource "aws_security_group" "pipeline_ecs_service_sg" {
...
+#   launch_type     = "FARGATE"
+
+#   network_configuration {
+#     subnets         = ["subnet-0497831b67192adc2",
+#                 "subnet-0acda1bd2efbf3922",
+#                 "subnet-0465f224c7432a02e"] 
+#     security_groups = [aws_security_group.pipeline_ecs_service_sg.id]
+#     assign_public_ip = true
+#   }
+# }
```
Comment (SurinaCS) [inline terraform/pipeline_ecs.tf:166]: do we need this commented out code?

File: terraform/upload_ecs.tf [hunk 7]
```diff
@@ -120,18 +120,18 @@ resource "aws_security_group" "raw_upload_ecs_service_sg" {
...
-  cluster         = data.aws_ecs_cluster.cluster.id
-  task_definition = aws_ecs_task_definition.c14_trendgineers_raw_upload.arn
-  desired_count   = 1
-  launch_type     = "FARGATE"
+# resource "aws_ecs_service" "service" {
+#   name            = "c14-trendgineers-raw-upload-service"
+#   cluster         = data.aws_ecs_cluster.cluster.id
+#   task_definition = aws_ecs_task_definition.c14_trendgineers_raw_upload.arn
+#   desired_count   = 1
+#   launch_type     = "FARGATE"
```
Comment (SurinaCS) [inline terraform/upload_ecs.tf:128]: same?

File: terraform/upload_ecs.tf [hunk 8]
```diff
@@ -120,18 +120,18 @@ resource "aws_security_group" "raw_upload_ecs_service_sg" {
...
-}
\ No newline at end of file
+#   network_configuration {
+#     subnets         = ["subnet-0497831b67192adc2",
+#                 "subnet-0acda1bd2efbf3922",
+#                 "subnet-0465f224c7432a02e"] 
+#     security_groups = [aws_security_group.raw_upload_ecs_service_sg.id]
+#     assign_public_ip = true
+#   }
+# }
```
Comment (SurinaCS) [inline terraform/upload_ecs.tf:137]: same?

**Review by SurinaCS — CHANGES_REQUESTED** [review]
great work! just a few queries on consistent returns and commented out code

File: rds_to_s3/etl_lambda.py (hunk 4 above)
Comment (Kurt812) [reply rds_to_s3/etl_lambda.py:104]: For some of these errors continuing with an empty dataframe is acceptable, however others represent failures that need explicit handling.

File: terraform/pipeline_ecs.tf (hunk 6 above)
Comment (Kurt812) [reply terraform/pipeline_ecs.tf:166]: We will need this code, however we don't want to create these services just yet, keeping them commented out to allow us to run terraform apply without worry

File: rds_to_s3/etl_lambda.py [hunk 9]
```diff
@@ -145,62 +129,73 @@ def upload_to_s3(bucket_name: str, file_name: str, object_name: str) -> None:
...
+            os.remove(tmp_file_name)
+            logging.info("Deleted local file: %s", tmp_file_name)
         else:
-            logging.warning("File %s does not exist.", file_name)
+            logging.warning("File %s does not exist.", tmp_file_name)
     except PermissionError as pe:
         logging.error(
-            "Permission denied while trying to delete %s: %s", file_name, pe)
+            "Permission denied while trying to delete %s: %s", tmp_file_name, pe)
         raise
```
Comment (SurinaCS) [inline rds_to_s3/etl_lambda.py:143]: great error handling here

**Review by SurinaCS — APPROVED** [review]
great work here!

**Pull request merged**